### PR TITLE
Introduce Off-Heap Technology for Large Arrays in a Region-Based Garbage Collector

### PR DESCRIPTION
**Pull request by jonoommen**
Introduce an optimized memory management for large arrays in a region-based garbage collector. More specifically, this is a new heap scheme and large object organization (in-heap and off-heap), an enhanced method for allocating large objects (proxy objects), and an optimized technique of accessing large arrays during runtime (JIT optimizations).

This new scheme not only unburdens large segregated array allocation from the heap, but it also preserves the illusion to the Just-In-Time (JIT) compiler through a clever object layout where all array elements can be accessed as if all arrays were contiguous (excluding zero-sized arrays). Furthermore, we can leverage such object layout (object proxies) to optimize memory management for large objects in a heap, by introducing off-heap.

Related to: https://github.com/eclipse-openj9/openj9/issues/14005

Signed-off-by: Jonathan Oommen [jon.oommen@gmail.com](mailto:jon.oommen@gmail.com)

### CODE REVIEW

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:146]: move down the declaration at the same line this is initialized.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:166]: provide a constructor or 2 for SparseDataTableEntr

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:289]: hm, wondering how accurate this will be over long run, and why is this needed anyway!?
but let's leave it for now.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:151]: extra brackets around NULL == entry

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:421]: indentation

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:161]: Think about some names that better describe what this accomplishes from a higher point of view (rather than talking about mechanics that occur within the method):
For example: mapSparseDataPtrToHeapObject. Then we would also have: findHeapObjectForSparseDataPtr etc.

File: gc/base/SparseVirtualMemory.cpp
Comment (amicic) [inline gc/base/SparseVirtualMemory.cpp:106]: A slighly shorter alternative to think about: allocateSparseFreeEntryAndMapToHeapObject

File: gc/base/GCExtensionsBase.hpp
Comment (amicic) [inline gc/base/GCExtensionsBase.hpp:576]: you can add something like:
large virtual memory for large objects, sparsely committed only for live ones

File: gc/base/SparseVirtualMemory.hpp
Comment (amicic) [inline gc/base/SparseVirtualMemory.hpp:46]: Give some description like:

Large virtual memory for allocation of large objects (their data portion). It's sparsely committed only for live objects (memory is eagerly committed/de-committed on allocate/free).
It also maintains mapping from data portion to proxy (heap) object.
The reverse mapping is not maintained (it's expected that heap object has an easy way to find data portion, for example, via pointer in the object header).

File: gc/base/SparseVirtualMemory.hpp
Comment (amicic) [inline gc/base/SparseVirtualMemory.hpp:67]: ITts ok to refer to OSX in the comment, but I'd add something like: this is primarily meant for OSX that is too lazy...

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (dmitripivkine) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:70]: Would you please define preprocessor constant for number of pool elements and pass it as parameter (instead of `6`)?

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (dmitripivkine) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:250]: please add space 
```suggestion
	while (NULL != current) {
```

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (dmitripivkine) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:243]: usually we declare `kill()` as `public` (as opposite to `newInstance()`) and `initialize()` and `tearDown()` as `protected` (to cover possible case in the future this class might have children need to call parent methods)

**Comment by dmitripivkine** [comment]
Seems copyright error for `configure` occur because this file does have different copyright

**Comment by dmitripivkine** [comment]
Would you please check each new file has empty line at the end? Usually we prefer to have <EOL> at the end of each line of the code, not <EOF> for last line

File: gc/base/SparseVirtualMemory.hpp
Comment (dmitripivkine) [inline gc/base/SparseVirtualMemory.hpp:72]: The same comment as above for `MM_SparseAddressOrderedFixedSizeDataPool`
- should `initialize()` and `tearDown()` be `protected`?
- should `kill()` be `public` and be called when we tear down class called `MM_SparseVirtualMemory::newInstance()`?

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (jonoommen) [reply gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:78]: No strong opinion for me either way. I can make the changes if it's decided this should be a class.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (jonoommen) [reply gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:241]: Yes, good point. Will make these changes.

File: gc/base/SparseVirtualMemory.hpp
Comment (jonoommen) [reply gc/base/SparseVirtualMemory.hpp:72]: Yes, thank you, good point. Will make these changes.

**Comment by jonoommen** [comment]
> Seems copyright error for `configure` occur because this file does have different copyright

I see. Looking at the previous times this file was modified, copyright changes were not made. Should this error be ignored for this file, or should I add copyright information to this file?

**Comment by dmitripivkine** [comment]
> > Seems copyright error for `configure` occur because this file does have different copyright
> 
> I see. Looking at the previous times this file was modified, copyright changes were not made. Should this error be ignored for this file, or should I add copyright information to this file?

I think this error can be ignored, final decision for Commiter obviously

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (amicic) [reply gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:78]: and to align with coding standards, prepend class fields with '_', and then the formal arguments dont need appended '_'

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:67]: identifier = NULL?

btw, couldn't the 3 above be just one, using default values for params?

File: gc/base/SparseVirtualMemory.cpp
Comment (amicic) [reply gc/base/SparseVirtualMemory.cpp:132]: You put thumb up, but did not change the name but you did change the name at another place with similar suggestion (the point of the change was mostly to use keyword 'Map' rather than 'AddData'. It's ok if like your original name more, but just making sure you did not forget, but (potentially) wanted to rename it.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (jonoommen) [reply gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:67]: Yes I will modify this, thank you. As for `identifier`, would you rather I just declared it instead of initializing it as well?

File: gc/base/SparseVirtualMemory.cpp
Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:147]: Ok 👍  Should I try removing this is and seeing if there are any stability issues?

Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:147]: Sounds good. Looking at my OpenJ9 changes at https://github.com/eclipse-openj9/openj9/pull/14667/files#diff-8a82e4e705ff1500c5a925ce893d948655c5d2964dbf17378190b43c3f1955a2, I explicitly call `commitMemory` in `MM_HeapRegionDataForAllocate::removeFromArrayletLeafList`, would this clear be necessary?

Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:147]: Ok, sounds good, we can delay this for now.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:47]: we don't use this style of intialization - move it to constructor(s)

Comment (jonoommen) [reply gc/base/SparseAddressOrderedFixedSizeDataPool.hpp]: Done 👍

File: gc/base/VirtualMemory.hpp
Comment (amicic) [inline gc/base/VirtualMemory.hpp:202]: i'd rather move some members from private to protected than make the subclass a friend (or something else is an issue?)

File: gc/base/SparseVirtualMemory.cpp
Comment (dmitripivkine) [inline gc/base/SparseVirtualMemory.cpp:76]: is it intentional to have float `2.0` in calculation operating with `uintptr_t` ?

Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:76]: There may have been a reason when I did it, but I don't believe it is necessary. I will update this.

File: gc/base/VirtualMemory.hpp
Comment (jonoommen) [reply gc/base/VirtualMemory.hpp:202]: Ok, will do 👍  No there is nothing else that should be an issue.

File: gc/base/SparseVirtualMemory.cpp
Comment (dmitripivkine) [inline gc/base/SparseVirtualMemory.cpp:143]: You can add comment before this block this is ultra safe decision, I am sure we can significantly reduce list of platforms required zeroing after commit

File: gc/base/SparseVirtualMemory.cpp
Comment (dmitripivkine) [inline gc/base/SparseVirtualMemory.cpp:79]: This assertion is useless - region size and page size are defined as power of two. However there is possibility that page size is larger than region size. Can logic handle this situation properly?

File: gc/base/SparseAddressOrderedFixedSizeDataPool.cpp
Comment (dmitripivkine) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.cpp:170]: Here and other similar places: Coding Standards recommend to avoid multi return functions. Would you please modify?

File: gc/base/SparseVirtualMemory.cpp
Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:79]: Logic can handle this for double mapping, but not currently for off-heap. I will add the same required conditions as well now for off-heap. Similar to double mapping, off-heap will now only be allowed to be enabled if the below condition is met: 

`if (!extensions->memoryManager->isLargePage(env, pagesize) || (pagesize <= extensions->getOmrVM()->_arrayletLeafSize))`

File: gc/base/SparseVirtualMemory.cpp
Comment (amicic) [inline gc/base/SparseVirtualMemory.cpp:91]: minor, but it could really be just one trace point with an extra string param (mapped from success bool)

File: gc/base/SparseVirtualMemory.hpp
Comment (jonoommen) [reply gc/base/SparseVirtualMemory.hpp:62]: Ok, noted 👍  And as you eluded to, the primary reason I didn't do any testing comparison for alternatives was because of the low contention.

File: gc/base/SparseAddressOrderedFixedSizeDataPool.hpp
Comment (amicic) [inline gc/base/SparseAddressOrderedFixedSizeDataPool.hpp:57]: break it in multiple lines (like constructor for MM_SparseHeapLinkedFreeHeader)

File: gc/base/SparseVirtualMemory.cpp
Comment (jonoommen) [reply gc/base/SparseVirtualMemory.cpp:91]: Ok 👍 The only other difference in these traces is that in `j9mm.tdf`, I have labelled `Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_failure` as a `TraceException` and `Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_success` as a `TraceEvent`. If with this new info the one trace statement is still preferred, I can definitely make this change.

Comment (amicic) [reply gc/base/SparseVirtualMemory.cpp:91]: Honestly, don't what difference it makes. Possibly for tools like HC to filter and show only Exception!?
Do you know @dmitripivkine ?

Comment (dmitripivkine) [reply gc/base/SparseVirtualMemory.cpp:91]: I am not sure we do have customers to use this. May be HC as you said. Do we have different default level for `TraceException`?  The main difference to have two or one tracepoint is possibility to enable/disable them individually. So using two different tracepoints make sense, at least for me. And "failed" case tracepoint might be enabled by default

Comment (amicic) [reply gc/base/SparseVirtualMemory.cpp:91]: ok, let's leave them separate

**Review by dmitripivkine — APPROVED** [review]

**Comment by jonoommen** [comment]
@0xdaryl Not urgent, but this change now has the needed approval to move forward with the change.

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by 0xdaryl** [comment]
Jenkins build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**
